File: 0020-valid-parentheses/0020-valid-parentheses.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let map: HashMap<char, char> = 
            [('(', ')'), ('[', ']'), ('{', '}')]
            .into_iter()
            .collect();

        let mut stack: Vec<char> = Vec::new();

        for ch in s.chars() {
            if map.contains_key(&ch) {
                stack.push(ch);
            } else {
                match stack.pop() {
                    Some(open) => {
                        if map.get(&open) != Some(&ch) {
                            return false;
                        }
                    }
                    None => return false,
                }
            }
        }
        stack.is_empty()
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        m = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        stack = []

        for c in s:
            if c in m.values():
                stack.append(c)
            else:
                if not stack or stack.pop() != m[c]:
                    return False

        return not stack
